# JCP Fall 26 — Review: Weeks 2 and 3

This notebook goes with the review slides. Every section uses the same three cell types:

1. **Predict**: read the cell and write your guess in the `# Guess:` comment *before* you run it.
2. **Fix**: the cell is broken on purpose. Read the error, find the bug, fix it.
3. **Your Turn**: replace every `...` with your own code.

Examples use **`tips`** (restaurant bills). Your Turn cells use **`titanic`** (passengers on the Titanic). Same skills, different data, so you write the code yourself instead of copying it.

Keep each cell short. If you're stuck for more than 2 minutes, ask your partner, then ask us.

In [139]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.metrics import mean_squared_error, confusion_matrix, accuracy_score, precision_score, recall_score

print("pandas version:", pd.__version__)

pandas version: 3.0.3


In [140]:
# We'll be using example datasets from Seaborn
tips = sns.load_dataset("tips")
titanic = sns.load_dataset("titanic")

---
## 0. Python Data Types and Dictionaries

**Predict:** what will each line print? Write your guesses first.

In [141]:
# Guess: int, float, str, bool, list?
print(type(25)) 
print(type(3.5))
print(type("25"))
print(type(True))


<class 'int'>
<class 'float'>
<class 'str'>
<class 'bool'>


**Example:** a dictionary is labeled pairs. Run each cell and watch what comes back.

In [142]:
menu = {
    "item": ["Latte", "Mocha", "Tea"],
    "price": [4.50, 5.25, 3.00]
}
menu["price"]

[4.5, 5.25, 3.0]

In [143]:
menu_df = pd.DataFrame(menu)
menu_df

,item,price
0,Latte,4.50
1,Mocha,5.25
2,Tea,3.00


In [144]:
# One column = Series. The whole table = DataFrame.
print(type(menu_df["price"]))
print(type(menu_df))

<class 'pandas.Series'>
<class 'pandas.DataFrame'>


In [145]:
menu_df.dtypes

item         str
price    float64
dtype: object

**Fix:** this cell has **two** bugs. Run it, fix the first error, run it again, fix the second.

In [148]:
broken = {
    "item": ["Latte", "Mocha", "Tea"],
    "price": [4.50, 5.25, 3.00]
}

pd.DataFrame(broken)

,item,price
0,Latte,4.50
1,Mocha,5.25
2,Tea,3.00


**Your Turn:** build a dictionary of 3 classes you're taking, with keys `"course"`, `"professor"`, and `"day"`. Turn it into a DataFrame and check its dtypes.

In [149]:
import pandas as pd
my_classes = {
    "course": ["Hist 160", "Envecon 100"],
    "professor": ["John", "Mark"],
    "day": ["Mon", "Tues"]
}
classes_df = pd.DataFrame(my_classes)
classes_df

,course,professor,day
0,Hist 160,John,Mon
1,Envecon 100,Mark,Tues


In [150]:
classes_df.dtypes

course       str
professor    str
day          str
dtype: object

---
## 1. First Look at a Dataset

**Example:** four quick checks on `tips`.

In [151]:
tips.shape

(244, 7)

In [152]:
tips.head()

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4


In [153]:
tips.dtypes

total_bill     float64
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
dtype: object

In [154]:
tips["day"].value_counts()

day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64

**Your Turn:** same four checks on `titanic`. How many rows? Which column tells you who survived? How many passengers were in each `class`?

In [155]:
titanic.shape

(891, 15)

In [156]:
titanic.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


In [157]:
titanic["survived"]

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: survived, Length: 891, dtype: int64

In [158]:
titanic["class"].value_counts()

class
Third     491
First     216
Second    184
Name: count, dtype: int64

---
## 2. Selecting and Filtering

**Predict:** for each cell, guess *Series or DataFrame*, or the exact value.

In [159]:
# Guess:
tips["tip"].head(3)

0    1.01
1    1.66
2    3.50
Name: tip, dtype: float64

In [160]:
# Guess:
tips[["tip"]].head(3)

,tip
0,1.01
1,1.66
2,3.50


In [161]:
# Guess:
tips.iloc[0, 1]

np.float64(1.01)

In [162]:
# Guess:
tips.loc[0, "day"]

'Sun'

**Example:** filtering in two steps. First the condition, then the brackets.

In [163]:
# Step 1: the condition is a column of True/False
(tips["size"] > 4).head()

0    False
1    False
2    False
3    False
4    False
Name: size, dtype: bool

In [164]:
# Step 2: put it inside brackets to keep only the True rows
tips[tips["size"] > 4]

,total_bill,tip,sex,smoker,day,time,size
125,29.80,4.20,Female,No,Thur,Lunch,6
141,34.30,6.70,Male,No,Thur,Lunch,6
142,41.19,5.00,Male,No,Thur,Lunch,5
143,27.05,5.00,Female,No,Thur,Lunch,6
155,29.85,5.14,Female,No,Sun,Dinner,5
156,48.17,5.00,Male,No,Sun,Dinner,6
185,20.69,5.00,Male,No,Sun,Dinner,5
187,30.46,2.00,Male,Yes,Sun,Dinner,5
216,28.15,3.00,Male,Yes,Sat,Dinner,5


In [165]:
# Two conditions: parentheses around EACH one
tips[(tips["size"] > 4) & (tips["day"] == "Sun")]

,total_bill,tip,sex,smoker,day,time,size
155,29.85,5.14,Female,No,Sun,Dinner,5
156,48.17,5.00,Male,No,Sun,Dinner,6
185,20.69,5.00,Male,No,Sun,Dinner,5
187,30.46,2.00,Male,Yes,Sun,Dinner,5


**Fix:** this should return Saturday tables with more than 4 people.

In [167]:
tips[(tips["size"] > 4) & (tips["day"] == "Sat")]

,total_bill,tip,sex,smoker,day,time,size
216,28.15,3.0,Male,Yes,Sat,Dinner,5


**Your Turn:** find all female passengers in 1st class (`pclass == 1`). How many are there?

In [168]:
first_class_women = titanic[(titanic["pclass"] == 1) & (titanic["sex"] == "female")]
len(first_class_women)

94

---
## 3. Mutability, Groupby, and Merge

**Predict:** after running the first cell, what will `tips.head(1)` show? The biggest tip ($10.00) or something else?

In [169]:
tips.sort_values("tip", ascending=False)

,total_bill,tip,sex,smoker,day,time,size
170,50.81,10.00,Male,Yes,Sat,Dinner,3
212,48.33,9.00,Male,No,Sat,Dinner,4
23,39.42,7.58,Male,No,Sat,Dinner,4
59,48.27,6.73,Male,No,Sat,Dinner,4
141,34.30,6.70,Male,No,Thur,Lunch,6
...,...,...,...,...,...,...,...
0,16.99,1.01,Female,No,Sun,Dinner,2
236,12.60,1.00,Male,Yes,Sat,Dinner,2
111,7.25,1.00,Female,No,Sat,Dinner,1
67,3.07,1.00,Female,Yes,Sat,Dinner,1


In [170]:
# Guess:
tips.head(1)

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2


In [171]:
tips_sorted=tips.sort_values("tip", ascending=False)
tips_sorted.head(1)

,total_bill,tip,sex,smoker,day,time,size
170,50.81,10.0,Male,Yes,Sat,Dinner,3


**Fix:** this should rename `total_bill` to `bill`, then show the new column.

In [172]:
tips = tips.rename(columns={"total_bill": "bill"})
tips["bill"].head()

0    16.99
1    10.34
2    21.01
3    23.68
4    24.59
Name: bill, dtype: float64

**Example:** groupby. Split by `day`, take the mean of `tip` in each group, combine.

In [173]:
tips.groupby("day")["tip"].mean()

day
Thur    2.771452
Fri     2.734737
Sat     2.993103
Sun     3.255132
Name: tip, dtype: float64

In [174]:
# More than one summary at once
tips.groupby("day")["tip"].agg(["mean", "count"])

,mean,count
day,,
Thur,2.771452,62
Fri,2.734737,19
Sat,2.993103,87
Sun,3.255132,76


**Example:** merge. Watch the shape before and after.

In [175]:
day_info = pd.DataFrame({
    "day": ["Thur", "Fri", "Sat", "Sun"],
    "weekend": [False, False, True, True]
})
day_info

,day,weekend
0,Thur,False
1,Fri,False
2,Sat,True
3,Sun,True


In [176]:
tips_weekend = pd.merge(tips, day_info, on="day")
print("before:", tips.shape)
print("after: ", tips_weekend.shape)
tips_weekend.head(3)

before: (244, 7)
after:  (244, 8)


,bill,tip,sex,smoker,day,time,size,weekend
0,16.99,1.01,Female,No,Sun,Dinner,2,True
1,10.34,1.66,Male,No,Sun,Dinner,3,True
2,21.01,3.50,Male,No,Sun,Dinner,3,True


**Your Turn:** what fraction of passengers survived in each `pclass`? (Hint: `survived` is 0 or 1, so its mean is the survival rate.)

In [177]:
titanic.groupby("pclass")["survived"].mean()

pclass
1    0.629630
2    0.472826
3    0.242363
Name: survived, dtype: float64

---
## 4. Missing Values

`tips` has no missing values, so we'll make a copy and blank out 3 tips on purpose.

In [178]:
tips_messy = tips.copy()
tips_messy.loc[[0, 5, 10], "tip"] = np.nan
tips_messy.isna().sum()

bill      0
tip       3
sex       0
smoker    0
day       0
time      0
size      0
dtype: int64

In [179]:
# Option 1: drop the rows
print("before:", tips_messy.shape)
print("after: ", tips_messy.dropna().shape)

before: (244, 7)
after:  (241, 7)


In [180]:
# Option 2: impute with the median
median_tip = tips_messy["tip"].median()
tips_messy["tip"].fillna(median_tip).isna().sum()

np.int64(0)

**Fix:** this should remove the missing rows, so the count at the end should be 0.

In [181]:
tips_messy.dropna()
tips_messy["tip"].isna().sum()

np.int64(3)

**Your Turn:** clean `titanic`.

1. Count missing values per column, then the *fraction* missing.
2. Make `titanic_clean`: drop the `deck` column, fill `age` with its median, then drop the rows still missing `embarked`.
3. Check that nothing is missing anymore.

In [182]:
titanic.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [183]:
titanic.isna().mean()

survived       0.000000
pclass         0.000000
sex            0.000000
age            0.198653
sibsp          0.000000
parch          0.000000
fare           0.000000
embarked       0.002245
class          0.000000
who            0.000000
adult_male     0.000000
deck           0.772166
embark_town    0.002245
alive          0.000000
alone          0.000000
dtype: float64

In [184]:
titanic_clean = titanic.drop(columns="deck")
titanic_clean["age"] = titanic_clean["age"].fillna(titanic_clean["age"].median())
titanic_clean = titanic_clean.dropna(subset=["embarked"])
titanic_clean.shape

(889, 14)

In [188]:
titanic_clean.isna().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

**Your Turn:** plot a histogram of `fare` from `titanic_clean`. It's heavily right-skewed, so add a `log_fare` column (use `np.log(... + 1)`, since some fares are 0) and plot that too. What changed?

In [189]:
titanic_clean = titanic.drop(columns=['deck'])
median_age = titanic_clean["age"].median()
titanic_clean["age"]=titanic_clean["age"].fillna(median_age)
titanic_clean=titanic_clean.dropna(subset=["embarked"])
titanic_clean.shape




(889, 14)

---
## 6. Regression: Predicting a Number

**Example:** predict `tip` from the bill, party size, and day. Run one cell at a time.

**Fix:** models only accept numbers. Run this, read the error, then move on to the next cell to see the fix.

In [190]:
LinearRegression().fit(tips[["total_bill", "day"]], tips["tip"])

KeyError: "['total_bill'] not in index"

In [201]:
# One-hot encode day: Thur is dropped and becomes the reference
tips_model = pd.get_dummies(tips, columns=["day"], drop_first=True, dtype=int)

tips_model[["bill", "size", "day_Fri", "day_Sat", "day_Sun"]].head()

,bill,size,day_Fri,day_Sat,day_Sun
0,16.99,2,0,0,1
1,10.34,3,0,0,1
2,21.01,3,0,0,1
3,23.68,2,0,0,1
4,24.59,4,0,0,1


In [203]:
X = tips_model[["bill", "size", "day_Fri", "day_Sat", "day_Sun"]]
y = tips_model["tip"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print("train:", X_train.shape, " test:", X_test.shape)

train: (195, 5)  test: (49, 5)


In [204]:
reg = LinearRegression()
reg.fit(X_train, y_train)
pd.Series(reg.coef_, index=X.columns).round(3)

bill       0.092
size       0.254
day_Fri    0.035
day_Sat   -0.130
day_Sun    0.044
dtype: float64

**Predict:** the model's test MSE vs. a constant model that always guesses the average tip. Which is lower?

In [205]:
model_mse = mean_squared_error(y_test, reg.predict(X_test))
baseline_mse = mean_squared_error(y_test, [y_train.mean()] * len(y_test))
print("model MSE:   ", round(model_mse, 3))
print("baseline MSE:", round(baseline_mse, 3))

model MSE:    0.666
baseline MSE: 1.449


**Your Turn:** predict a passenger's `fare` (a number) from `titanic_clean`.

1. One-hot encode `embarked` (drop first, `dtype=int`). Check the new column names.
2. Use features `age`, `pclass`, `sibsp`, and the `embarked` columns.
3. Split 80/20 with `random_state=42`, fit a `LinearRegression`, and look at the coefficients.
4. Compare your test MSE to the baseline that always guesses the average fare.

In [206]:
titanic_reg = pd.get_dummies(titanic_clean, columns=["embarked"], drop_first=True, dtype=int)
titanic_reg[["age", "pclass", "sibsp", "embarked_Q"]].head()

,age,pclass,sibsp,embarked_Q
0,22.0,3,1,0
1,38.0,1,1,0
2,26.0,3,0,0
3,35.0,1,1,0
4,35.0,3,0,0


In [207]:
X_f = titanic_reg[["age", "pclass", "sibsp", "embarked_Q"]]
y_f = titanic_reg["fare"]

X_train, X_test, y_train, y_test = train_test_split(
    X_f, y_f, test_size=0.2, random_state=42
)

print("train:", X_train.shape, "test:", X_test.shape)

train: (711, 4) test: (178, 4)


In [208]:
fare_reg = LinearRegression()
fare_reg.fit(X_train, y_train)

pd.Series(fare_reg.coef_, index=X_f.columns).round(3)

age           -0.240
pclass       -35.033
sibsp          8.989
embarked_Q     3.141
dtype: float64

In [209]:
fare_mse = mean_squared_error(y_test, fare_reg.predict(X_test))
fare_baseline = mean_squared_error(y_test, [y_train.mean()] * len(y_test))
print("model MSE:   ", round(fare_mse, 3))
print("baseline MSE:", round(fare_baseline, 3))

model MSE:    1822.056
baseline MSE: 2667.977


*What does the `pclass` coefficient mean, in your words?* (Hint: `pclass` goes up by 1 when a passenger moves from 1st toward 3rd class.) 

The pclass coefficient is -35.033, meaning that for each one-level increase in pclass (moving from 1st toward 3rd class), the predicted fare decreases by about $35, holding the other variables constant.

---
## Done: Push Your Work

Run these in your terminal (not in this notebook):

```
git add Week_Review.ipynb
git commit -m "review notebook"
git push origin YOUR-BRANCH-NAME
```